bluesky_kafka/__init__.py

```diff
@@ -1,4 +1,287 @@
-__version__ = get_versions()['version']
+
+__version__ = get_versions()["version"]
 del get_versions
+
+logger = logging.getLogger(name="bluesky.kafka")
+
+
+def delivery_report(err, msg):
+    """
+    Called once for each message produced to indicate delivery result.
+    Triggered by poll() or flush().
+
+    Parameters
+    ----------
+    err
+    msg
+
+    """
+    if err is not None:
+        logger.error("message delivery failed: %s", err)
+    else:
+        logger.debug(
+            "message delivered to topic %s [partition %s]",
+            msg.topic(),
+            msg.partition(),
+        )
+
+
+class Publisher:
+    """
+    A callback that publishes documents to a Kafka server.
+
+    Reference: https://github.com/confluentinc/confluent-kafka-python/issues/137
+
+    There is no default configuration. A reasonable production configuration for use
+    with bluesky is Kafka's "idempotent" configuration specified by
+        producer_config = {
+            "enable.idempotence": True
+        }
+    This is short for
+        producer_config = {
+            "acks": "all",                              # acknowledge only after all brokers receive a message
+            "retries": sys.maxsize,                     # retry indefinitely
+            "max.in.flight.requests.per.connection": 5  # maintain message order *when retrying*
+        }
+
+    This means three things:
+        1) delivery acknowledgement is not sent until all replicate brokers have received a message
+        2) message delivery will be retried indefinitely (messages will not be dropped by the Producer)
+        3) message order will be maintained during retries
+
+    A reasonable testing configuration is
+        producer_config={
+            "acks": 1,
+            "request.timeout.ms": 5000,
+        }
+
+    Parameters
+    ----------
+    topic: str
+        Topic to which all messages will be published.
+    bootstrap_servers: str
+        Comma-delimited list of Kafka server addresses as a string such as ``'127.0.0.1:9092'``
+    key: str
+        Kafka "key" string. Specify a key to maintain message order. If None is specified
+        no ordering will be imposed on messages.
+    producer_config: dict, optional
+        Dictionary configuration information used to construct the underlying Kafka Producer
+    serializer: function, optional
+        Function to serialize data. Default is pickle.dumps.
+
+    Example
+    -------
+
+    Publish documents from a RunEngine to a Kafka broker on localhost on port 9092.
+
+    >>> publisher = Publisher(
+    >>>     topic="bluesky.documents",
+    >>>     bootstrap_servers='localhost:9092',
+    >>>     key="abcdef"
+    >>> )
+    >>> RE = RunEngine({})
+    >>> RE.subscribe(publisher)
+    """
+
+    def __init__(
+        self,
+        topic,
+        bootstrap_servers,
+        key,
+        producer_config=None,
+        serializer=pickle.dumps,
+    ):
+        self.topic = topic
```

Comment:
Do we want these to be mutable after `__init__` time, or should we hide them in read-only properties?

Reply:
The should be read-only.

Reply:
Done!